bpf: Coccinelle scripts for align_stack_8 and const qualifier

pchaigno opened the pull request:
Please see commit messages for details.

I'm planning to enable this in CI as a followup, if @qmonnet doesn't get to it first.

pchaigno committed:
bpf/Makefile: Introduce coccicheck to run all cocci patches

This commit introduces a new coccicheck target to be able to run all
Coccinelle patches independently from other checks (e.g., sparse).

That target doesn't apply the patch or even display the diff, but it will
print a message for each required change. If the cocci patch supports
adding the required changes automatically, it will also print a final
message with the command to apply all changes.

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
coccinelle: Find missing __align_stack_8 annotations

This commit adds a new Coccinelle script to find missing __align_stack_8
annotations, for on-stack objects of size > 8 bytes for which 8 bytes
isn't the natural object alignment.

Please note this script requires Coccinelle 1.0.7 to handle our #pragma
unroll (otherwise, some missing __align_stack_8 may be missed).

This semantic patch currently only signals missing __align_stack_8, but
we will be able to add them automatically once Coccinelle has support
for matching on attributes [1].

1 - https://systeme.lip6.fr/pipermail/cocci/2020-April/007309.html
Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
bpf: Prepare for const qualifiers patch

Just a couple manual changes that are needed for the semantic patch in
the following commit to apply and compile successfully.

Signed-off-by: Paul Chaignon <paul@cilium.io>

borkmann commented on bpf/lib/nat46.h:
```diff
@@ -221,7 +221,8 @@ static __always_inline int ipv6_prefix_match(struct in6_addr *addr,
 static __always_inline int ipv4_to_ipv6(struct __ctx_buff *ctx, struct iphdr *ip4,
-					int nh_off, union v6addr *v6_dst)
+					int nh_off,
+					const union v6addr *v6_dst)
```
small nit: could go to prior line

borkmann commented on bpf/include/bpf/ctx/xdp.h:
```diff
@@ -136,7 +136,8 @@ __csum_replace_by_4(__sum16 *sum, __wsum from, __wsum to)
-l3_csum_replace(struct xdp_md *ctx, __u64 off, const __u32 from, __u32 to,
+l3_csum_replace(const struct xdp_md *ctx, __u64 off, const __u32 from,
+		__u32 to,
 		__u32 flags)
```
small nit: could go to prior line

pchaigno replied:
:thinking: Not sure what happened here. Coccinelle is supposed to be good at handling this sort of things.

borkmann reviewed (approved):
Nice work, I like it! :-)

pchaigno committed:
coccinelle: Add const qualifiers to function parameters

This commit adds a new Coccinelle script to add const qualifiers to the
appropriate function parameters.

All other changes in bpf/ in this commit have been generated by that new
script.

Signed-off-by: Paul Chaignon <paul@cilium.io>

tklauser reviewed (approved):
Not too familiar with SmPL itself, so only briefly skimmed through. But the generated changes and the Makefile rule LGTM.

qmonnet reviewed (approved):
Here's your Janitors' review anyway. Awesome work, thanks! Let's merge it.
